### ISSUE TITLE
Retry?

### ISSUE THREAD

**Issue by dovy** [mharrisb1/cube-http-client#5]
Do you have any support for retries built in?

**Comment by mharrisb1** [comment]
Yes-ish. There is `max_retries` options in `ClientOptions`. This uses [`httpx.HTTPTransport`](https://www.python-httpx.org/advanced/transports/#http-transport). Unfortunately, this just handles simple retries and there is no way to specify more granular retry logic like for specific status codes.

https://github.com/mharrisb1/cube-http-client/blob/defd1c97623b584bb7eb70d233896e680ef330f5/src/cube_http/types/_client.py#L19

Example:

```py
import cube_http

cube = cube_http.Client({
    "url": "...",
    "token": "...",
    "max_retries": 3,
})
```

## Proposals

### Automatic retry on `/v1/load` response status code 100

The Rust client crate that I took inspiration from implements retries explicitly for `/v1/load` to retry on status code 100 responses.

https://github.com/cube-js/cube/blob/e1911d4ff4c754d7d7e0d3f2bd6116d7531004c8/rust/cubesql/cubeclient/src/apis/default_api.rs#L71

We could definitely do something similar here.

### Custom Transport

Create a custom retry transport and use that to set `http_client` in client options. This is the one that I'd like to do the most since it's the more general solution. I'd probably package this under the name `httpx-retry`.

It would involve creating a new Transport type:

```py
import httpx


class RetryTransport(httpx.BaseTransport):
    ...
```

Here, I'd have some way for the user to specify some retry logic like for [exponential backoff](https://en.wikipedia.org/wiki/Exponential_backoff) or any other policy. It would need to be generic enough to create any retry policy but also simple enough that the user shouldn't need to code an exponential backoff algorithm by hand.

What I envision:

```py
import cube_http
import httpx
from httpx_retry import RetryTransport

cube = cube.Client({
    "url": "...",
    "token": "...",
    "http_client": httpx.Client(transport=RetryTransport(...)),  # not sure what the class interface would be yet
})
```

**Comment by mharrisb1** [comment]
Retries should be defined using external lib [httpx-retry](https://pypi.org/project/httpx-retry). That way we keep this lib minimal and clean.

Example usage:

```py
import cube_http

import httpx
from httpx_retry import HTTPRetryTransport, RetryPolicy

exponential_retry = (
    RetryPolicy()
      .with_attempts(3)
      .with_min_delay(100)
      .with_multiplier(2)
)

cube = cube.Client({
    "url": "...",
    "token": "...",
    "http_client": httpx.Client(transport=HTTPRetryTransport(policy=exponential_retry)),
})
```
